From 810b40e1f5a04602d9f49ef0645997ec684ee197 Mon Sep 17 00:00:00 2001
From: Abhinav Kishan Potharaju <quic_apothara@quicinc.com>
Date: Wed, 16 Oct 2024 14:48:35 +0530
Subject: [PATCH] Updating 901-debloat_sock_diag.patch

Debloat: add some debloat patches, strip down procfs and make
O_DIRECT support optional, saves ~15K after lzma on MIPS

Change-Id: Ie352957cde973aae21749b4ff893a56be85bfaf2
Signed-off-by: Abhinav Kishan Potharaju <quic_apothara@quicinc.com>
---
 net/core/sock.c | 19 +++++++++++--------
 net/xdp/Kconfig |  1 +
 2 files changed, 12 insertions(+), 8 deletions(-)

diff --git a/net/core/sock.c b/net/core/sock.c
index ce5a00aa4030..73879be824ba 100644
--- a/net/core/sock.c
+++ b/net/core/sock.c
@@ -175,14 +175,17 @@ EXPORT_SYMBOL(sk_ns_capable);
 
 u64 __sock_gen_cookie(struct sock *sk)
 {
-       while (1) {
-               u64 res = atomic64_read(&sk->sk_cookie);
-
-               if (res)
-                       return res;
-               res = gen_cookie_next(&sock_cookie);
-               atomic64_cmpxchg(&sk->sk_cookie, 0, res);
-       }
+	u64 res = atomic64_read(&sk->sk_cookie);
+
+	if (!res) {
+		u64 new = gen_cookie_next(&sock_cookie);
+
+		atomic64_cmpxchg(&sk->sk_cookie, res, new);
+
+		/* Another thread might have changed sk_cookie before us. */
+		res = atomic64_read(&sk->sk_cookie);
+	}
+	return res;
 }
 
 /**
diff --git a/net/xdp/Kconfig b/net/xdp/Kconfig
index 71af2febe72a..719df49761a8 100644
--- a/net/xdp/Kconfig
+++ b/net/xdp/Kconfig
@@ -10,6 +10,7 @@ config XDP_SOCKETS
 config XDP_SOCKETS_DIAG
 	tristate "XDP sockets: monitoring interface"
 	depends on XDP_SOCKETS
+	select SOCK_DIAG
 	default n
 	help
 	  Support for PF_XDP sockets monitoring interface used by the ss tool.
-- 
2.34.1

